### ISSUE TITLE
🎓Fractional usage and calendaring

### ISSUE THREAD

**Issue by micahalcorn** [OriginProtocol/origin-js#146]
The eventual goal is to support fractional-use DApps that will involved a calendaring function. Some ideas around this are outlined in the [white paper](https://www.originprotocol.com/en/whitepaper).

**Things to Consider**
- mapping user-friendly times and smart contracts
- challenges with daylight savings, time zones, etc
- business process requirements
- [iCalendar standard](https://tools.ietf.org/html/rfc5545)
- variable pricing on-chain in an efficient format

**To Do**
- [x] UI changes supporting time intervals ranging from minutes to year(s)

**Comment by DanielVF** [comment]
One way that is flexible, simple, and gas cheap:

We could have N times slots. Each timeslot has a start and and time in UTC seconds (unix timestamp). The listing IPFS json has an array of these slots. The listing contract has an availability (bytes or array of numbers) as a bit string. We can do cheap bitwise operations on these to check availability or mark as used.

**Comment by joshfraser** [comment]
I love the idea of utilizing bit masks.

A couple things to remember, we'll eventually want the ability to set variable pricing for different time slots (we've been calling them "intervals") and we'll need some way to track who the purchaser of each Interval was, not just availability (although maybe we get that already from the purchase contract?).

It may also simplify things if each Interval is a fixed length of time. That allows you to have simple rules like every 6th and 7th interval should have a higher price (say for weekend pricing).

**Comment by DanielVF** [comment]
Are there any changes to the purchase flow that we need for this fractional usage sprint?

**Comment by nick** [comment]
I found some resources that may or may not be useful for this discussion:

- [Interval Tree implementation in Solidity](https://github.com/gnidan/interval-trees-solidity)
- [Timely Resource smart contract design](https://docs.google.com/document/d/11WCubnkejEs-1K_I17DgQSlNtGJO8TS2vXkTCijytGg/edit)

**Comment by joshfraser** [comment]
Thanks for kicking this off Daniel and enumerating the various decisions we need to make.

While different types of listings should be able to have variable interval sizes (or slots as you call them), I think we should have a fixed interval size for a specific listing. Things like being able to set a minimum number of intervals (ie. min 2 nights booking) is pretty important and easier to reason about when the intervals are a consistent size.

**Comment by wanderingstan** [comment]
@nick that Timely doc is interesting, though I can't find any evidence that they proposed an ERC. Or any evidence of who wrote it! Do you recall where you found it? Looks like it _might_ be from this company: https://www.gettimely.com/about/ ? 

I also came upon interval trees in my research! We have some extra "inside information" that makes our use case maybe simpler than the purely academic: we'll never insert intervals before "now", we know we'll never allow intervals to overlap, and we won't delete intervals.  

Still digging into this. Fun problem. 

@DanielVF where is the doc you shared in the eng call yesterday with the bitmask breakdown? I keep forgetting to bookmark it!

**Comment by nick** [comment]
I found the Timely Resource article [here](https://www.reddit.com/r/ethdev/comments/837ku1/proposed_erc_for_timebased_resources_feedback/) on reddit.

One idea I had that may help simplify our implementation of this is to offload the pricing and availability aspects to be off-chain, and simply have an initial booking be 'pending' until 'accepted' by the seller. This would allow very complex rules for availability and pricing without having to verify everything on-chain. This would also allow us to support buyers making offers below the official price, as users frequently do on Airbnb.

The contract could just keep an 2 arrays of iCal strings... one for available slots and one for bookings. It'd be up to the seller, or the dapp the seller is using, to check the requested slot is actually available before they go ahead and accept the listing. Perhaps we could incentivize good behavior by having the seller stake some tokens which they would lose if they accept an overlapping booking, don't respond in a timely manner, etc.

For 'instant' bookings, perhaps the seller could delegate accepting valid bookings to the bridge server or other 3rd party service, which would take on the responsibility of validating bookings against listings.

In short, I think it'd be difficult to get all the functionality we desire on-chain, perhaps we can store the 'bare bones' data on the blockchain without trying to validate availability and pricing at transaction time.

**Comment by cuongdo** [comment]
As @wanderingstan mentioned, we shouldn't need interval trees for non-overlapping intervals. A balanced BST is sufficient for storing those, as mentioned in the "[Naive approach" section of the Wikipedia article for interval trees](https://en.wikipedia.org/wiki/Interval_tree#Naive_approach). However, the cost of storing all the references required for a BST on the blockchain would likely make this approach expensive.

Using (potentially sparse) bitmasks seems appealing for its space savings. I'm going to think on this more.

@nick: Thanks for the links and for sharing your thoughts. In terms of moving the availability and pricing off-chain, I wouldn't feel comfortable putting the complexity to the seller or each DApp maker. However, I wonder how much responsibility could be shifted to origin.js.

**Comment by DanielVF** [comment]
Unrelated to the previous conversation:

Our hard work here is all about coming up with cheaper ways of representing this data - but we don't have to worry about it being possible or not. This all childishly simple if the seller pays, for example, five cents a timeslot.

**Comment by nick** [comment]
Yea I think origin.js can certainly provide helper methods to ease the setup and management of fractional usage listings. My main concern is how much data we should be putting on the blockchain instead of in IPFS, and whether the contract should be responsible for validation or not. 

For example, if a user submits a booking request to my listing contract to book my condo next July 13th-15th for 0.5ETH... should the contract validate that those slots are available and that the submitted price is correct? Or should we just let the seller (or some automated 3rd party) decide if they want to accept the booking manually. I believe the latter would provide us with much more flexibility as well as less contract complexity as we no longer need to represent slots and state in a way that can be validated in contract code.

We could just store the availability and pricing information in IPFS, then have a 'bytes' variable that gets appended to every time there is a new booking. For example: `J1-10 A3-4 S21-24` could mean a property is booked July 1st-10th, August 3rd-4th and September 21st-24th. (this is just a super simple example off the top of my head... not suggesting we actually use that format).

Representing slots on-chain could be expensive. If I'm a restaurant wanting to represent 50 tables where slots run 5pm-10pm at 15 minute intervals and accept bookings 3 months in advance, then that's 50 tables * 5 hours * 4 slots per hour * 90 days = 90,000 slots. A hotel with 500 rooms taking bookings 6 months in advance is also 90,000 slots. Represented as bits would be 90,000 / 256 = 350 words... a lot of storage. I think just storing actual booked slots would reduce the storage needed, at the cost of not being able to do validation (without significant complexity) on-chain.

**Comment by joshfraser** [comment]
I believe it's important to have on-chain validation. We add a lot of friction if every transaction has to be manually approved. And just like you don't want to count on client-side validation in traditional web apps, we can't trust client-side validation here either (people can always go around origin.js and talk to our smart contracts directly).

A few things we can do to reduce costs:

- Bitmasks for more efficient storage
- Sparsely populated arrays (don't pre-pay for storage). Have the buyers pay for their share of the storage costs instead of having the seller cover everything upfront 
- Use recurring rules and ranges to support variable pricing without touching each slot. Imagine a simplistic rules engine that can handle simple rules like: 
  - Normal slot costs 1 ETH
  - Every 5th or 6th slot costs 2 ETH
  - Slots 100 through 200 cost 3 ETH

**Comment by cuongdo** [comment]
For bitmasks, it's also possible to use simple forms of compression that work well with bitmaps, such as run-length encoding. Hopefully, the gas cost of compression/decompression for a well-picked algorithm is far lower than the storage cost of the uncompressed data. This would have to be prototyped and profiled.

**Comment by nick** [comment]
Totally agree that ideally everything would be validated on chain and that should be our ultimate goal. However, doing so will require complex data structures and validation code which may be able to handle the simplest cases but I suspect will begin to break down as we reveal more and more corner cases for different models.

I suspect in the real world (at least in the early days) booking requests will need to be validated by the seller anyway: 

- Home rentals are likely to have their property listed on Airbnb (and other services), as well as Origin. Unless we have some kind of syncing mechanism, bypassing origin.js won't help secure a guaranteed booking... the seller will need to confirm there are no conflicts and mark the Airbnb calendar as unavailable before confirming a blockchain booking. As I recall, initially all Airbnb listings had to be manually confirmed by the seller before they introduced 'instant bookings'.
- Restaurant bookings will need to integrate with their POS system and sync to the blockchain if they are to guarantee a table reservation without a confirmation step
- Sellers in general will likely want to personally validate buyers before confirming bookings. Personally I can't see myself guaranteeing a booking from anyone on the internet via a blockchain transaction without first double checking the buyer is legit manually, at least until I have complete trust in the system.

I think this differs from trusting client-side validation as ultimately the seller is in control of whether they accept a booking or not. There's no way for the buyer to trick the seller... the seller would just reject the booking after reviewing the request.

That being said, I don't see why we shouldn't at least try an implementation and see how far we get :-)

p.s. one other consideration I thought of is that sellers may not want 'price paid' for past transactions traceable. If you regularly make special deals for certain clients and those prices are public, it may make negotiation difficult. Eg "I saw you rented out your place last month for $100/nt but it's currently listed for $200. Can you give me the same deal?". Or "You charged your last client $50/hr, I want that price too".

**Comment by wanderingstan** [comment]
I couldn't sleep last night (think I'm getting a cold 😷) but was thinking about interval time representation and linked lists. Sorry for long rambling post. 

# On chain datetime representation

We don't need to be hesitant about having a way to get to "real" datetime values on the blockchain. That is, even if we are storing things as just slot numbers, we can and should provide ways to translate these to real timestamps when they are returned to the dapp. 

Recall that reads from the blockchain are free. *And* it's free to call functions, no matter how complicated, as long as they don't alter the EVM state.

So we can have functions convert our slot numbers into real datetimes using libraries like these: 
https://github.com/pipermerriam/ethereum-datetime
https://github.com/bokkypoobah/BokkyPooBahsDateTimeLibrary

I think it's important that there be no ambiguity about what the contracts represent. I.e. the interpretation of slots to date/times should not happen (only) off-chain. 


# Ordered Linked LIsts

In a traditional ordered linked-list, the `insert()` function does two things:

1. Scan the list to find the right place to insert new node
2. Update node pointers to actually do the insert

In our blockchain world, we can massively cut on gas costs by splitting these steps apart. 

Recall that reading from the chain is free. (Also free is to call functions, no matter how complicated, as long as they don't alter the EVM state.) So we can do step 1 without paying gas. 

Then we just have to have a modified `insert()` function that also accepts where you want to make the insert, _and is able to verify that it is a valid place to make the insertion_. The good news in our case is that it's trivial to verify whether a desired insertion is valid. 

We then put the onus on the dapp to do the "research" about where a booking belongs in the linked list. (Possibly assisted by indexing server, though I doubt that's necesarry). 

In prose: 

A traditional insert call would be, "Hey Mr. Linked-list! Please insert a booking for Xavier from interval 16-18."
I'm proposing the insert call to be, "Hey Mr. Linked-list! Please insert a booking for Xavier from interval 16-18, I'm sure you'll agree that there is an opening right after the existing booking for Charlie."


### Bookings on number line
```
 1 
 2 ▲ Alice [2,4]
 3 | 
 4 ▼
 5 
 6 
 7 
 8 ▲ Bob [8,9]
 9 ▼
10 ▲ Charlie [10,12]
11 |
12 ▼
13 
14 
15 
16 
17 
18 
19 ▲ Diane [19,20]
20 ▼
```

## As Linked list
```
HEAD: →A

{id: A, interval: [ 2, 4], next: →B    }
{id: B, interval: [ 8, 9], next: →C    }
{id: C, interval: [10,12], next: →D    }
{id: D, interval: [19,20], next: →null }
```

## Inserting

This is the crux: 

We pass along the booking that our booking should come after. In other words, we do the "work" of figuring out where in the list it should go *before* calling `insert`. 

Then `insert` only has to *verify* that it's in the right place by confirming that our new booking falls between the `after` booking and the current `next` of `after`.)

```
insert( after: C, {id: X, interval: [16,18]} )
```

`insert` confirms that our new booking `X` does indeed come after `C` and before `D` -- thus a valid booking!  

Rough solidity pseudo-code:
```
function insert(node after_node, node new_node) {

  // Verify we're inserting in right place with no overlap
  require(new_node.interval.start > after_node.interval.end, 
    "New node must start after after_node's end");
  require(new_node.interval.end < after_node.next.interval.start,
    "New node must end before after_node.next's start")

  // Now do regular linked list insert stuff
  new_node.next = after_node.next
  after_node.next = new_node
}
```

### Bookings numberline view after insert
```
 1 
 2 ▲ Alice [2,4]
 3 | 
 4 ▼
 5 
 6 
 7 
 8 ▲ Bob [8,9]
 9 ▼
10 ▲ Charlie [10,12]
11 |
12 ▼
13 
14 
15 
16 ▲ *Xavier [16,18]
17 |
18 ▼
19 ▲ Diane [19,20]
20 ▼
```

## Additional concerns



### Simultaneous overlapping bookings

The approach should still work cases where multiple `insert()`s are called within one block, such that only one of them can "win". It seems acceptable behavior that one of the bookings get rejected. This would be the same for any representation where there is a race to make a booking.

For example, if Yvonne was trying to make a booking ar roughly the same time as Xavier (i.e. so that their `insert()` calls happened in the same block) , but her booking overlaps with his. 

```
insert( after: C, {id: Y, interval: [15,17]} )
```
In this case, it would simply be up to the miner as to who's booking "wins". I don't think we can do much about that. 

### Simultaneous non-overlapping bookings

The simple approach would need a little more inteligence to deal with cases where multiple *non-overlapping* `insert()`s are called within one block. 

Imagine that Zane is trying to make a booking ar roughly the same time as Xavier, such that they're in the same block:
```
insert( after: C, {id: Z, interval: [14,15]} )
```
If Zane's insert is executed by the miner first, then Xavier's insert will fail: the `insert()` function will see Zane as being the next booking after Charlie, and there is no room for Xavier's booking there!

Possible solutions: if the "blocking" booking ends before our new booking begins, treat that one as our `next` and repeat. I.e. the code should see that Zane's booking ends before Xavier's begins, so it's okay to look a little further.  

_Note that this essentially means that the insert() code is thus beginning to do a traditional list scan--the thing we wanted to avoid._ I think that this is okay: its up to the submitted to determin how much gas they're willing to pay to cover these rare situations. Additonally, we could limit this scan to only scan bookings that were made in the current block (so it doesn't somehow runaway and start scannig whole list.)


### Tracking "Now"

`HEAD` points to the start of the linked list. For our purposes, it would be useful to also maintain a pointer to the *first booking which is not in the past*. Call it `CURRENT`. This will save future callers from having to start at the beginning of the list every time they're scanning. 

This could be accomplished by requiring an `insert()` caller to also pass along the node that is either (a) currently in progress (ie we're in middle of a booking) or (b) the node that is most recently in the past. (ie "now" lies in the gap between this old booking and the `next` of this node.) 

To allow for inaccuracies in blockchain's version of `now`, this could be softened to require only +/- 10 minutes or whatever.

**Comment by DanielVF** [comment]
One possible rule based storage system:

![image](https://user-images.githubusercontent.com/837/40859599-5ce1ae04-65b0-11e8-9160-6daf2811bc97.png)

I think these are the minimum needed slot functions for the listing contract:

- isSlotAvailable(slotId) returns bool
- slotPrice(slotId) returns price
- purchaseBefore(slotId) returns timestamp
- completedAfter(slotId) returns timestamp

**Comment by cuongdo** [comment]
I like @DanielVF's proposal for calendar events for which the resource is either available or unavailable. However, it doesn't address the [events/ticketing scenario](https://docs.google.com/document/d/1_UuT49_UUpCtKhLL-MBlMFu8Sg0XPRP3jGIAJzFM3KU/edit). Specifically, each event is both scheduled and has a fixed quantity of tickets. So, we need more bits somewhere to represent that.

The other case not addressed is the [freelancing scenario](qT4EkfO3qxI2rB4xr3g1T0ZY-pyNojcQ2FuprLK2U/edit) where a service needs to be done within a certain number of days. It's likely that a freelancer / contractor would be able to take on multiple jobs in any given period. Unfortunately, this makes some types of freelancing a form of overlapping scheduling, which hasn't been considered in-scope for this effort.

**Comment by wanderingstan** [comment]
@cuongdo, the events/ticketing scenario doesn't seem to actually be a case of fractional usage, I think. That is, there is no "thing" that people use/possess at different times.

Having 200 tickets for sale to a Taylor Swift concert seems more akin to having 200 widgets for sale on eBay, than to an Airbnb listing. (With obvious difference that tickets are only worth anything up until the event, and the possibility of different pricing based on seat location, time until show, etc)

**Comment by cuongdo** [comment]
@wanderingstan You're right that events & ticketing aren't fractional usage. As mentioned in the [fractional usage/calendaring scenarios doc](https://docs.google.com/document/d/1_UuT49_UUpCtKhLL-MBlMFu8Sg0XPRP3jGIAJzFM3KU/edit) (search for "event dates"), there is the potential for ticketed events to recur. It's possible to push recurring events into a different layer of the stack to keep things simpler for fractional usage.

**Comment by wanderingstan** [comment]
In effort to get clear in my own head, I did a first draft of mapping out a "Taxonomy of Listings" to better delineate the major listing types that we may support:
https://docs.google.com/document/d/1RH-bEnUp7EmhtFOUHc5pofvqbVsS8pnmS4_ZpH0SsKY/edit#

@cuongdo has made the good point that we need to get clear on whats in/out of scope, and I think it'd be good to get things in context of a larger roadmap.

**Comment by coleman415** [comment]
Quick note for everyone, the fractional usage/calendaring scenarios doc comparing existing/traditional sharing economy products referenced by @cuongdo has been been broken up into different docs for each use case (e.g. homesharing, freelancing, ticketing). You can find them in this [folder](https://drive.google.com/open?id=14inftkQMtpTXESaqn_pN0XMJrb11xxEs).

Please let me know if there are changes you want made to the format/content of these docs.

**Comment by ambertch** [comment]
@wanderingstan mentioned something about 'what data is required for arbitration scenarios' - I know that with TrueBit state channels, the idea is to have a verification game in which enough data is available on-chain to allow anyone to verify an off-chain computation that was performed (also reference something like the Lightning Network, where the parties have to agree in order to commit transaction results back to the main chain).

In this case, the data that is required to verify a calculation might be inputs/outputs, plus a set of rules that the application logic should follow.

This makes life more complex for dapp developers, since what it amounts to is telling them "you can implement certain pieces of the application logic offchain, but in addition to posting the results back to the chain you also need to use a definition scheme/language and define a set of rules that can be used to validate whatever you do offchain" although it has the advantage, as I think @nick alludes to, that smart contract logic can be kept more general and have less complex data structures / validation code.

**Comment by ambertch** [comment]
As an example, in addition to weekend pricing Airbnb also supports weekly and monthly rates:

<img width="315" alt="screenshot 2018-06-06 13 41 06" src="https://user-images.githubusercontent.com/156263/41074841-d766bc98-69be-11e8-89ef-4d8d8f0c5e40.png">
<img width="321" alt="screenshot 2018-06-06 13 42 42" src="https://user-images.githubusercontent.com/156263/41074843-dac47d1c-69be-11e8-96e8-19ffd17dc58a.png">

In this case, if all of the application logic were contained in the smart contract, there would need to be params for flags corresponding to whether there is X, Y, Z type discounts, as well as the multiplier of the discount. What if instead of a percentage, the discount should be a flat rate? 

If a dapp developer does the pricing calculations, they might write serialized key-value pairs that define a ruleset for validation, ex:
`{ weekendMultiplier: 1.2, priceIncreaseDays: { <timestampThanksgiving>: 50, <timestampChristmas>: 60, priceDescreaseDays: { <timestampWhenZombiesInvadeTown: 99.99 } }, weeklyDiscountPercentage: 20}`

For all validation logic existing on chain, I wonder if there is a pattern analogous to pluggable middleware, like Express middleware in node.js. That way contracts containing core application logic don't need to be upgraded whenever new business requirements are discovered (although versioning and release notes address that as well).

**Comment by nick** [comment]
The way I see it, validation with no confirmation step is all or nothing... anything that requires validation from the seller renders validation on-chain unnecessary. If the seller needs to confirm any detail of the transaction, they may as well validate and confirm everything else off-chain too.

I think it'll be difficult to convince Airbnb users to use Origin unless we support the same functionality as Airbnb. Validating everything on-chain (just for the Airbnb case) would require checking:

**User credentials**
- ER725 checks, eg Origin validated, Govt ID

**Availability**
- All requested slots are available
- Number of slots requested meets criteria (min, max)
- Advanced notice (eg must be at least 3 days before start date)
- Preparation time (auto-block 1 or 2 nights before and after each booking)
- Booking window (cannot book more than x months in advance)
  
**Price**
- Base price per slot (should allow overrides for specific slots, eg Christmas)
- Weekend pricing
- Weekly / monthly discount
- One off fees (eg cleaning)
- Other extra charges (charge per extra guest after x guests, late checkout)
- Occupancy taxes and fees (can be complex, eg new tax only applies after certain date)
- Security deposit
  
All of these variables would have to be represented in a smart contract, and the validation function would need to come in under the gas limit. And this is just for Airbnb... all the other use cases probably have their own special requirements we'd also need to consider.

Instead of doing availability and price validation on-chain, we can do so in origin-js instead. The smart contract would then only be responsible for validating user credentials and storing a list of active bookings, resulting in much simpler contract code that would work for all the use cases we have outlined.

Doing price and availability validation off-chain will also make writing custom rules easier for partners - which I'm sure will be an important factor for them in deciding whether Origin is a good fit for them or not.

Although doing price and availability validation on-chain would be ideal, I think at this stage it's too big of an undertaking given the myriad of other challenges and timing issues we face. Writing all the price/availability validation off-chain in origin-js will probably take several weeks, if not a month... trying to do it all on-chain and get it right plus all the auditing of complex pricing functions will likely take a lot longer.

**Comment by cuongdo** [comment]
@nick Could you elaborate on your suggestion? Specifically, if enforcing availability and price is done by `origin-js` instead of on-chain, what on-chain checks would occur to prevent bad/invalid bookings from being made by code that makes contract calls *without* `origin-js`? If the listing allows automatic booking without seller approval, what prevents a bad actor from booking with a price that's too low for the given dates if the contract code can't validate the price?

**Comment by nick** [comment]
There would be nothing to stop bad/invalid bookings from being made at the contract level. However I would argue that _guaranteeing_ a booking this way is impractical for our purposes anyway (unless users use Origin exclusively) due to potential conflicts with other services. For example if someone lists their home on both Origin and Airbnb, and separate buyers book on both Origin and Airbnb at the same time. The 'guaranteed' booking would have to be cancelled on either Origin or Airbnb (and I'm pretty sure there are penalties for cancelling on Airbnb).

The way I see this working (using home sharing as an example) is as follows:

1. Seller posts a listing in the same way as today. All pricing and availability variables are stored as JSON in IPFS. The listing contract stores only a list of booked slots.
2. origin-js takes this JSON as input and provides convenience methods for showing pricing and calendar information for dapps. Booked slots (taken from the contract state) are also marked as unavailable.
3. origin-js provides a 'buy' method which validates the listing is available for the given dates and calculates the correct price. The booking request is sent to the blockchain with a status of 'pending approval'. A corresponding booking JSON object is stored in IPFS (encrypted with seller's public key?) with stuff like number of guests, special requests, etc.
4. Seller is notified of booking request. origin-js provides convenience methods for dapp to validate booking request given the pricing/availability rules specified in the listing JSON. This would prevent buyers submitting invalid booking JSON directly to IPFS and bypassing origin-js. Seller reviews request and approves.
5. Disputes would be resolved the same way as with any other listing, via arbitration

Essentially the on-chain responsibilities would be limited to user validation (ERC725), escrow, arbitration and cancelations (eg return half the fee if cancelation is within 7 days). Even just that functionality I think would take a fair amount of gas.

**Advantages**
- Minimal changes to our existing contracts
- No special contract logic for different use cases
- Complex pricing/availability models all in origin-js
- Support buyers submitting a request with a lower price (eg offering $100/nt instead of $150)

**Disadvantage** No guaranteed instant booking

**Comment by wanderingstan** [comment]
### Non-instantbook

@nick is slowly winning me over to the non-instantbook camp. 
- Being simpler to understand and to code, it's a good use-case to build first.
- Impossible (or very complex) to cross-sync with AirBnB bookings
- It is the use-case that AirBnB began with.
- We can figure out instant-book later, using what we learn. 
- Interesting: There is nothing stopping a DApp/service from implementing their own instant book. I.e. service watches for booking requests, if one comes in with good attestations & price, it automatically accepts it on behalf of the seller.

And to highlight something that wasn't immediately obvious to me: In this scenario, the price-per-slot information is really just "suggested price" price information--the seller is free to accept or decline any offer. 

Looking through the other [Fractional Listing use-cases](https://docs.google.com/document/d/1RH-bEnUp7EmhtFOUHc5pofvqbVsS8pnmS4_ZpH0SsKY/edit#heading=h.4eo1v679eivx), the one that would most need instant-book IMO would be car-sharing (like Getaround). It would also need location, however, which we haven't begun to address.
 
### Instantbook thoughts

OTOH, when we do instantbook (as we eventually must), I think the logic should be a separate module from the actual booking/purchase. That is, the resulting `Purchase` contract should look the same whether it came from a manually approved booking or an instantbook. The instantbook logic is simply an on-chain version of the off-chain service I described above, encoding the conditions under which the seller will accept a booking automatically. This neat segmentation would serve us well going forward.

**Comment by tyleryasaka** [comment]
To add my 2 cents, I also tend agree with @nick 's approach. I think that whatever we do, we're going to get it wrong the first time. So rather than spending a tremendous amount of time and energy trying to build the ideal solution for these more complex use cases, it would be better to build a very minimal solution, put it out in the wild, and collect feedback on that and improve from there. In this specific scenario, I think that means not trying to support instant book right now, for the reasons @nick and @wanderingstan have listed. I think the aspect of launching a new platform and making it trustworthy could make it difficult for us to get instant book working in a usable way for a long time. (Seems like it will be a lot easier to get people to initially adopt Origin if they can use it alongside AirBnB and other services, as @wanderingstan and @nick have noted.)

That said - I do think we can possibly build out an experimental instant book option into our contract. For example, I like the idea of allowing sellers, upon creating a listing, to be able to point to an "instant book" contract for the listing, which would have a function that would return a boolean for whether or not a purchase is allowed, for a given slot id and ERC725 identity; and maybe a pricing function that would return a price for a given slot id. We could build out one or two specific instant book contracts for specific use cases, without trying to build out a generalized solution that will work for everything. (We might eventually get to a generalized solution, but I think we will need to experiment with more specific use cases before we have the knowledge we need to properly generalize.)

This way we could at least try out the idea of instant book for very simple use cases without being dependent on it. Of course that would require some development time so there would be a tradeoff. But I guess what I'm saying is, it doesn't have to be all or nothing. We can build out non-instant booking as our primary use case, and then experiment with opt-in instant booking on the side.

**Comment by tyleryasaka** [comment]
Re-reading @wanderingstan 's last comment, I think we are thinking along the same lines as far as separating the concern of purchase approval from everything else. We could have a purchase stage for `awaiting approval`, and perhaps a method on the purchase contract called `isAutoApproved`. If the listing contract has its `instantBookContract` property set with a contract address, then `isAutoApproved` will call that contract's `isApproved` method and return its result. If the `instantBookContract` property is not set, then `isAutoApproved` will return false, and the purchase contract will remain in the `awaiting approval` stage until the seller calls the purchase contract's `approve` method. Either way, once the purchase contract leaves the `awaiting approval` stage to the next stage, everything else would function *exactly* the same either way.

If we structure our contracts this way, then I think we will have full flexibility going forward. We can do any of the following:
- Experiment with specific instant book contracts that we want the dapp to support
- Just support non-instant bookings to begin with
- Some combination of the two

As long as we keep the logic cleanly separated, we should be able to pick one strategy now but not be locked into it if we ever need to pivot in the future (which seems plausible if not likely).

**Comment by matthewliu** [comment]
From a purely product perspective, I don't see Instant Book as a must have requirement for v1. It would be great to have of course.

Instant book makes the most sense for ecommerce transactions (multiple quantity), some instances of hotels/homesharing, and instances of events (concerts, yoga class slots).

In the real world:
- Craigslist style sales require seller approval (vs. Amazon or Shopify which automatically fulfills)
- Airbnb listings can be either instant book or seller-approved (vs. hotels which are always instant booked)
- Home services (cleaning, repairs), beauty services (on-demand massage, makeup appointment), contractor services (do I take on this design project or marketing project), all require seller approval
- (Out of scope) As real-time as Uber and Postmates are, they also require seller approval. The routing and acceptance process is just super fast.

I would say the majority of our starting use cases can do without instant book, even if they are not optimal. As long as we don't peg ourselves into a corner that prevents this in the future, I'm fine pushing this out.

As for engineering implementation, most of you are much better informed than I am on the tradeoffs.

**Comment by tyleryasaka** [comment]
One last comment! During my last job where we were building a marketplace for recycled auto parts, we initially built the system where everything was automatically sold based off of the availability we had in our system. However, this had numerous problems, and we had a very high cancellation rate of our orders, which was very frustrating for buyers. The problem was that the sellers were using our system in addition to their other systems (because we were a fledgling marketplace and they were opting into our service as just an additional source of revenue, not the primary source). Their inventory would be sold from one of their other systems or over the counter, or they would find that the listing was incorrect and the part was actually not in working condition. This was in spite of a very complex system we had in place to sync with their existing inventory management systems.

For a major partner that we integrated into our system, one of their requirements was that we build an approval process for them, so that they would have a chance to manually review an order before it went through. So despite all of our efforts to build a fully automated system, we ended up building a manual approval process anyway.

In addition to the availability problems I just described, we had trust problems as well. Our sellers were manually approved into the system on a case-by-case basis for us, so that's how we ensured that the seller was trustworthy. But we also had buyer trust issues, and spent a lot of time dealing with fraud and returns. If we had this much trouble building a trustworthy marketplace in a centralized system, I can only imagine it will be more difficult in a decentralized system. All that to say, it seems like it will be a lot easier to build a trustworthy system around manual rather than automatic approvals, at least as a first iteration.

I don't know how much this experience applies to the use cases we're considering (I assume used auto parts is pretty low on our priority list). But it's just a personal experience I have with how difficult it is to build an automated approval system that *just works*.

**Comment by franckc** [comment]
Just want to share my experience on a past project I worked on that is in some sense related to this discussion.

At PayPal, my team implemented the incentive engine responsible for calculating how much Points, Gift Certificates and Coupons could be used by a user at checkout time to pay for eBay items. The rules were stored as XML blobs in a database table. The rule engine would take as input those rules and would output the list of incentives that could be used and associated amounts (ex: on this item, user can use 100 points and $20 GC and gets a 20% off by using coupon XX). The transactions recorded in the "ledger" database table would just record a pointer to the rule config that was used (e.g. a rule ID) and the amount of each incentive used. Having a configurable and editable set of rules that were decoupled from the transaction ledger gave us a lot of flexibility. 

So overall, +1 for starting simple by storing the availability and pricing rules off-chain and using the chain mostly as a transaction ledger.

Just my 2 cents :)

**Comment by cuongdo** [comment]
OK, I believe we have a consensus that starting with approval-required listings and using off-chain storage (IPFS) for pricing & availability is the technically more feasible approach for 1.0. This allows us to use a subset of iCal/jCal to serialize the recurrence rules, which means we can save ourselves engineering time by using existing, tested JS libraries to parse the rules.

I'd also like to explore @tyleryasaka's idea to delegate approvals to external contracts. I'm curious what the potential security/abuse issues would be for that, and I'm sure some care will need to be exercised.

Unless I hear objections, I'll work on the design doc for approval-required fractional usage listings.

**Comment by tyleryasaka** [comment]
@cuongdo Yeah, I'll be glad to provide more details on the approach I was thinking. Probably will be easiest to do that once you have a first version of your design doc.

**Comment by cuongdo** [comment]
While there are still some incomplete parts, the design doc is ready for review: https://docs.google.com/document/d/131XzwVNGlZ6itINS_Gy0o0dYzRtN_5NQXbcqXZuL2vk/edit?usp=sharing

@joshfraser @wanderingstan @DanielVF: given the amount of thinking you've already done, I'm looking to you all for the bulk of the reviewing

@tyleryasaka: Please take a look, as you're prototyping the Origin.js part of this

**Comment by nick** [comment]
I think the confirmation step should be optional for regular listings too and not just fractional listings. For example I may want to offer something for sale only to a particular group of people (eg a discounted item to members of some organization), so an optional confirmation step could also be useful for regular listings.

I think we should also try to be as generic as possible with terminology, for example 'request' instead of 'reserve' and just 'CONFIRMED' instead of 'RESERVATION_CONFIRMED'.

Regarding timezones, datetimes and reservation formats... how about we represent a slot as a single uint256:

4 bytes for slot format, 8 bytes for timezone, 10 bytes for start time, 10 bytes for end time.

If we represent in separate start, end and timezone variables that would take 3 times more space and offer no flexibility if we wish to offer alternative formats later on. Slot format '1' might mean '8 bytes for timezone, 10 bytes for start time, 10 bytes for end time' whereas slot format '2' could mean '14 bytes for start time, 14 bytes for end time'.

Do we need a 'seller rejected' phase? Since rejecting a listing costs gas, what incentive does the seller have to reject a request? Perhaps we should instead just rely on a timeout value (specified when the listing is created)? The seller can communicate that the listing is rejected (eg price should be higher, or slot is unavailable) off-chain via a channel specified in the reservation request IPFS JSON (eg IPFS pubsub, Telegram chat, etc). That channel would also be used for general communication between buyer/seller.

Regarding storing timezone and dates in IPFS: we could just use the JavaScript date object, which includes timezone information? E.g: 

```
new Date().toString()
"Fri Jun 15 2018 10:25:02 GMT+0100 (British Summer Time)"
```

This will make it easier to parse in JS without a library, e.g.:

```
Number(new Date("Fri Jun 15 2018 10:25:02 GMT+0100 (British Summer Time)"))
1529054702000
```

We could also consider the Schema.org DateTime type: https://schema.org/DateTime

Regarding payments... I think the value in ETH (or other ERC20 token) should be sent along with the reservation request. If the seller does not accept within the specified time period, the buyer is automatically refunded. If the seller accepts the listing, the value goes into escrow until the booking is considered a success (eg x days after the booking ends, unless a dispute is initiated).

Overall looking good though, this is fun stuff to think about 😄

**Comment by cuongdo** [comment]
Thanks for reviewing the doc, @wanderingstan!

It seems that at a minimum, we want weekend/weekday pricing for home sharing-listings. What would availability look like in this iCal format you're talking about? Would it look like a series of alternating blocks of 5 weekdays and 2 weekends? Or would it be two RRULES, one for weekdays and one for weekends? 

While I agree that showing availability is a convenience, I think that it's a useful one. I don't recall the last time I went to a reservation site (Airbnb, my barber, restaurant reservation) with no indication if availability. I think that would undesirable for UX.

**Comment by cuongdo** [comment]
One idea for IPFS blob versioning is to include the previous hash in each IPFS blob. So, you'd be able to go back in time from the current hash.

**Comment by nick** [comment]
One thing that worries me about iCal is the lack of good JS library support. It seems the only real option is [ical.js](https://github.com/mozilla-comm/ical.js/) which, looking at the [commit history](https://github.com/mozilla-comm/ical.js/commits/master), doesn't get a lot of development. The [API](http://mozilla-comm.github.io/ical.js/api/) doesn't look all that friendly either, and from looking at it briefly I'm not sure supports all out use cases easily. I think we should at least investigate the library further, as well as other implementations (eg Python for the bridge server) to make sure we're not painting ourselves into a corner by making it a part of our protocol. I agree it can be useful for import/export for calendar apps, but I'm not sure of the benefits beyond that.

As an alternative, we can use a combination of objects from Schema.org (not to be confused with JSON Schema, which is for object validation). For our purposes, the following schemas could be useful:

- [Offer](https://schema.org/Offer). Example fields: availabilityStarts, availabilityEnds, deliveryLeadTime, eligibleRegion
- [DateTime](https://schema.org/DateTime). Example fields: startDate, endDate, checkinTime
- [LodgingReservation](https://schema.org/LodgingReservation). Example fields: numAdults, numChildren
- [Accommodation](https://schema.org/Accommodation). Example fields: floorSize, numberOfRooms
- [PriceSpecification](https://schema.org/PriceSpecification). Example fields: priceCurrency, minPrice, validFrom

The full list of Schema.org types is [here](https://schema.org/docs/full.html)

[Here](https://schema.org/docs/hotels.html) is a good example of using Schema.org objects to model a Hotel

Regarding availability, I think we should also think about having an option to store this information off-chain. The IPFS object could have an optional URL property where availability can be stored. Since availability is merely a 'suggestion' as Stan points out, having it 'verifiably correct' on IPFS and on-chain is less important. Having it optionally stored off-chain will allow users to update their calendar or subscribe to an Airbnb sync service (for example) without having to perform an on-chain update.

If you're a 'superhost' managing hundreds of properties and need to submit a blockchain transaction every time you need to update a reservation... it could become prohibitive.

**Comment by wanderingstan** [comment]
@cuongdo , I know you're out this week. Hope you don't check github too often. :)

> 1. It seems that at a minimum, we want weekend/weekday pricing for home sharing-listings. What would availability look like in this iCal format you're talking about? Would it look like a series of alternating blocks of 5 weekdays and 2 weekends? Or would it be two RRULES, one for weekdays and one for weekends?

The pricing could be done as two RRULES, or by just listing out different prices for weekends for the foreseeable future. As @DanielVF has pointed out, we probably want relatively short-lived contracts, so having even a few hundred entries (for about a year's worth of weekends) isn't bad. 

But the key point is that pricing is "just suggestions" to be displayed in the UI—we don't have to decide the absolute right way: One DApp could use RRules while another one just creates a ton of event entries. The end goal is that other DApps can consume this pricing data and display it on a calendar. 

> 2. While I agree that showing availability is a convenience, I think that it's a useful one. I don't recall the last time I went to a reservation site (Airbnb, my barber, restaurant reservation) with no indication of availability. I think that would undesirable for UX.

Sorry, I didn't communicate well: I meant availability is a "convenience" in the sense that it is not required for the final on-chain money-exchanges-hands Purchase contract. (I.e. It's no contractually binding and can't play any role in arbitration.) When we show an availability calendar to the user, we are implicitly saying "These are dates that the seller has indicated are available, but we have no way of confirming it, and this is in no way binding."  We **definitely** want our DApp to show this to users, but ultimately the decision of book/no-book is up to the seller. 

> 3. One idea for IPFS blob versioning is to include the previous hash in each IPFS blob. So, you'd be able to go back in time from the current hash.

We will still need some sort of on-chain versioning: Otherwise, a seller could update the contract to point to a maliciously updated IPFS blob, which could include the hash of still another maliciously updated IPFS blob. E.g. I'm being taken to arbitration after the buyer points out that my house actually doesn't have a hot tub. I update my IPFS blob to remove the hot tub part. And if need be, I can create a fake "previous" IPFS blob for it to point to. 

> 4. Good observation about not needing to store availability twice. For UX, it'd be good to allow the seller to block off days where the property is not available. It'd be strange if they had to make a Purchase to make that happen, so there should be another mechanism for doing that.

I'm realizing that we actually have 2 sorts of availability:
- **Internal**: The availability that we (i.e. the Origin contracts) know about: Purchases/Bookings made on our system. These have start/end dates where we can [probably] presume listing is unavailable. This data is on-chain (and can be read at no cost). I'm still undecided if we want to _enforce_ unavailability from this, but we should definitely _show_ it. 
- **External**: Availability indicated by the seller, either manual (by seller marking off dates) or automated (as AirBnB and VRBO do with exported iCal URLs). We all agree there's no reason to put this on-chain, as it can change on a whim and is not contractually binding. I suspect this could just be a URI that points to an iCal feed somewhere. For manual entries, the DApp could create an iCal IPFS blob. (Not sure about importing an external AirBnB feed. @joshfraser is rightfully allergic to having the DApp access non-IPFS URIs, as this would enable user tracking. Maybe the DApp fetches it, and merges it into the user's manually-entered blackout dates?)

**Comment by wanderingstan** [comment]
@nick : You've got me torn! iCal totally sucks, but its the de-facto standard for availability-calendar syncing (AirBnB, VRBO, and the rest.) Seems we'd need to support it at least as far as (1) having the DApp able to read an AirBnB/VRBO-generated iCal of existing bookings to mark as unavailable and (2) being able to generate a similar iCal that can be read by AirBnB/VRBO so they can mark off Origin-booked dates. Maybe this could just have this deed derived from the schema.org data? (The good news is none of this needs to happen on-chain at least!)

BTW [LodgingReservation](https://schema.org/LodgingReservation) seems like exactly something that could/should be stored in the Purchase contract for homesharing.

**Pricing vs Availability** : My thought was: Given that we need to support iCal for availability, it makes sense to also use it for pricing. Could be in a combined iCal file, or one for pricing and one for availability. But IMO from a purely data point of view, "July 4-5, 2018 is unavailable" is not far from of "July 4-5, 2018 is $121/night". So I'd suggest we use the same parsing/logic to process both. 

...But maybe the schema.org stuff is certainly more comprehensive. @nick, can you elaborate on how you imagine them being used? I'm roughly thinking, the `Listing` would have IPFS which contains JSON schema with info about the home (or whatever resource), this json contains IPFS hash of another file which contains the availability/pricing information. But this is all getting pretty convoluted! I might start up some more diagrams. :)

**Comment by nick** [comment]
I think we can support iCal for import/export/sync without necessarily using it as the 'source of truth' in IPFS. Instead, we can define our own format for storing availability + pricing, then have toiCal / from iCal methods that marshal the data to and from the iCal format when we need to import/export/sync. This will give us more flexibility and allow us to write our own API without being constrained by the iCal format at the data storage level.

Does anyone here have an active Airbnb listing with upcoming booking so we can look at the iCal export to see what it supports? If it doesn't show pricing information, that will have to be inserted by the user manually when importing from Airbnb. Here are the instructions for [exporting iCal](https://www.airbnb.co.uk/help/article/99/how-do-i-sync-my-airbnb-calendar-with-another-calendar) from Airbnb.

**Comment by wanderingstan** [comment]
@nick [This is a recent AirBnB iCal ](https://gateway.originprotocol.com/ipfs/QmW8AjzcneqUKykFFM6czjBJM7EWi8sxGoi5yaLvjfq9s1)

Looks like whoever made the "standard" followed iCal's lead with having ALLCAPPS KEYS followed by data. :)

Sample data for one iCal event:
```
NIGHTS: 2
PHONE:  +1 (XXX) XXX-XXXX
EMAIL: (no email alias available)
PROPERTY: 2BD near Moscone, Union Square & BART
```

**Comment by wanderingstan** [comment]
> we can define our own format for storing availability + pricing, then have toiCal / from iCal methods

I like this approach, but we'd need some iCal file at a URL that persists, which I'm assuming would be an IPFS blob on our gateway. (Like the iCal link in previous comment). And we'd like this iCal to be discoverable (by other Dapps, indexers, etc) so there would need to be some link to it from our Listing IPFS or contract.

**Comment by nick** [comment]
Sure, exporting to a separate IPFS blob would work well I think... then we can have a pointer to that in the main listing. We can also consider leveraging IPNS so we can update availability and pricing without having to update the IPFS hash on-chain.

**Comment by tyleryasaka** [comment]
This is a long thread!

I've been tasked with putting together a working prototype for fractional usage contracts. I've read through @cuongdo 's design doc as well as everyone's comments and considered the different ideas proposed.

My goal is to follow agile principles and keep this first iteration as simple as possible. I think the discussion here is great in the sense that we're thinking things through and making sure we've considered many angles before we start building. That said, I don't think we need to (or can) solve all of these detailed problems in the first iteration. I think we need to get something simple but functional out there and experiment with it.

With that in mind, this is what I'm thinking for the first prototype:

1. I'm going to keep both pricing *and* availability off-chain for now. I'll structure the contracts so that we will be able to easily support instant booking contracts in the future, but won't worry much about that now. My reasoning for doing availability off-chain right now is that it's simpler, and not necessary for a functional prototype. I am not necessarily opposed to storing availability data on-chain (haven't really thought this through enough to have an opinion), but I do think that the cost/benefit ratio for doing that in an initial prototype is not sufficient to justify it. (I'll take great care to structure the contracts so that we can easily add on-chain availability whenever we decide to do that.)
1. I will support versioning for the IFPS hash because that seems like an essential feature (and is trivial to implement), as @wanderingstan has discussed.
1. iCal debate: no opinion yet. I will get a first prototype of the contracts in place first as iCal stuff will be completely off-chain right now - then I will revisit the iCal debate.
1. Purchase contract structure (see below)

## Purchase contract proposal

After experimenting with the solidity contracts, I'm convinced that our best course of action (for now at least) is to have a single purchase contract for both types of listings. What this means is that we'll delegate some logic back to the listing contract (such as checking that payment is sufficient). It also means that we will want to simplify the names of our stages so that they're more generic. We basically want all of our listing purchases to go through the same states, even if the state changes get triggered in different ways for different types of listings, or if certain stages are skipped altogether for certain types.

I'm proposing the following stages, for this first iteration:
- `AWAITING_PAYMENT` (initial state)
- `AWAITING_SELLER_APPROVAL` (buyer has requested purchase and sent payment)
- `SELLER_REJECTED` (seller rejected buyer request)
- `IN_ESCROW` (payment is being held in escrow, awaiting some event or action)
- `BUYER_PENDING`
- `SELLER_PENDING`
- `IN_DISPUTE`
- `REVIEW_PERIOD`
- `COMPLETE`

For our current listings, the state machine looks like:
- `AWAITING_PAYMENT`
- buyer makes purchase -> `IN_ESCROW`
- seller marks as shipped -> `BUYER_PENDING`
- (rest the same as before)

For our fractional usage listings:
- `AWAITING_PAYMENT`
- buyer makes purchase -> `AWAITING_SELLER_APPROVAL`
- seller accepts -> `IN_ESCROW`
- reservation end time passes (or marked as such by seller if using off-chain availability) -> `BUYER_PENDING`
- (rest the same as before)

I've already made a good amount of progress on a branch for this prototype: https://github.com/originprotocol/origin-js/tree/tyleryasaka/fractional-usage

**Comment by wanderingstan** [comment]
This thread is long but rich! I'm learning a ton and agree that these are fun, interesting problems!

@tyleryasaka Can't wait to try your branch. What would be a good entry-point for trying it out? Looking at the smart contracts? 

## schema.org+ jsonschema

...getting a little more general than just Fractional: 

It crystalized for me on the flight that what we probably want is to replace the schemas @joshfraser admitted to making in a half hour last year (😀) with the schema.org schemas @nick found (Like [LodgingReservation](https://schema.org/LodgingReservation) ), **but** we'd need to encode them in JSON schema. Presumably in JSON-LD, schema.org's [preferred json format](https://schema.org/docs/datamodel.html).   

A little rummaging on the web led me to what [seems to be the definitive thread on the subject of JSON-LD + JSON-schema](https://github.com/json-schema-org/json-schema-spec/issues/309).  But I'll freely admit that I don't fully grok the discussion—can't even tell if an answer was reached! Is this what we'd want to do? 

(I'm 90% sure that we want to use JSON schema. But is there any argument for me made for say, using "plain" JSON-LD and writing our own form-renderer? I predict that we will outgrow react-jsonschema-form pretty soon, as aure designs kick-ass UI/UX flows that can't be captured by a generic renderer.)

**Comment by nick** [comment]
I think schema.org and jsonschema can work well together... schema.org for defining things like field names and structure and jsonschema for doing the actual object validation. We may outgrow react-jsonschema-form but we can still leverage [ajv](https://github.com/epoberezkin/ajv), for example, to do the heavy lifting of validating an object.

Here's an example of a LodgingReservation as JSON-LD:

```
{
  "@context": "http://schema.org",
  "@type": "LodgingReservation",
  "reservationId": "abc456",
  "reservationStatus": "http://schema.org/ReservationConfirmed",
  "underName": {
    "@type": "Person",
    "name": "John Smith"
  },
  "reservationFor": {
    "@type": "LodgingBusiness",
    "name": "Hilton San Francisco Union Square",
    "address": {
      "@type": "PostalAddress",
      "streetAddress": "333 O'Farrell St",
      "addressLocality": "San Francisco",
      "addressRegion": "CA",
      "postalCode": "94102",
      "addressCountry": "US"
    },
    "telephone": "415-771-1400"
  },
  "checkinTime": "2017-04-11T16:00:00-08:00",
  "checkoutTime": "2017-04-13T11:00:00-08:00"
}
```

**Comment by tyleryasaka** [comment]
Opened a pull request for v1 of fractional usage contracts: https://github.com/OriginProtocol/origin-js/pull/251

**Comment by lucadgtrip** [comment]
I’m adding my perspective as a Dapp marketplace project in the Homesharing space.

INSTANT BOOKING: NOT NEEDED NOW
The Instant Booking feature wouldn’t even be used by Hosts at the beginning.
Origin Dapps will start as a minor booking provider for them.
If an Host gets 10% of bookings from Origin and 90% from Airbnb, she will fear Airbnb and won’t care too much about Origin, which will be just an experiment.
This leads to cancellations (in case of Instant Bookings) or refusals (in case of On Request Bookings).
Most Hosts will try Origin out and manually review each booking.
We’ve seen this happening over the years with Airbnb and Booking.com competitors such as Wimdu, HouseTrip or Tripadvisor: most Hosts would not opt-in for Instant Bookings or often refuse them when they got one (due to not updated prices for instance). 
A refused Instant Booking is toxic for the Guests’ experience and the platform.
Just imagine booking a place, buy a flight and then get a cancellation: you now need to find another place for the same dates and price.
So, I agree with the choice.

Of course if Instant Booking was trivial to develop and could be given as an option, some Hosts would use it, which brings us to another questions: would Guests use it? Probably not.
Trust is needed before users feel confident to Instant Book on a platform.
Even more so on a blockchain/crypto platform.
 
Having said that: in the long run it WILL be needed, as it makes the user experience so much better and all OTAs push hard for it.
Have you seen “Daddy's Home Two” where Mel Gibson (instant) books a house on Airbnb while the others are fighting over what to do on Xmas? That shows how important they are to Airbnb.
Also, Instant Booking carries another message: “price and availability are 100% correct”, which boosts also non Instant Bookings and in general the platform credibility.
But, again, it’s too early for that.

ICAL: 100% NEEDED NOW
Yes, iCal is needed (for availabilities).
One of the first questions every Hosts asks when they hear about a new platform is “is there iCal?”.
The reason is again, that for them it’s an experiment, and they won’t keep availabilities updated for sure. This leads to Guests having to contact dozens of Hosts and getting “sorry, not available” replies, while dates on calendar were available.
This is extremely dangerous and it’s still today a problem for Airbnb.
Availability information is one of the few things Airbnb shares.
They’re not really happy to do that, but they probably realised that not all Hosts accept to be exclusive with them and in order to attract more Hosts, they added the feature.
Booking.com is still fighting this, it goes completely against their silos mentality but they were also forced to slowly, unwillingly, give up this data.
Expedia still does not have iCal and is seeing little traction in the vacation rental space, in  part due to this. No Host right in his mind will list in Expedia (which is Instant Booking Only) without iCal. This is maybe why they spent 3.9 Billion USD to acquire Homeaway.
So, let’s definetely use this data they are forced to share.

AVAILABILITIES ARE GOLD
It’s early for this kind of reasoning but let’s realise that ultimately the fight amongst Platform is won by those who have the most availabilities. 
If Origin just gets a few nights per year because most dates are already booked by Booking and Airbnb, there’s not much we can do.
I’ve got some ideas on how to fight this battle but this may not be the place or the time, let me know if you’d like to know more.

I apologize if this was unrelated to dev but I think that the underlying business dynamics are important.
Again, let me know if I’m out of place or if this should go in Discord.

**Comment by wanderingstan** [comment]
Thanks @lucadgtrip , this is the right place. 

Sounds like on iCal you are 100% that we be able to _import_ iCal from AirBnB (and others).  How important do you think it is that we _export_ iCal for other services to ingest? (I think difficulty-wise it's about the same, but just for our prioritization.)

**Dapp Developers -  Issues with importing iCal** 

For importing iCal data, do you think we can get away with each Dapp pinging the URL (eg the AirBnB availability URL) on it's own? If tons of people start using Origin dapps, it could look like a DDOS attack. 

Last week @nick rightly chastised me for sharing the content of Josh's AirBnB iCal, because it contained guests' names and phone numbers. If these iCal URL's are stored in IPFS (or on chain), then this is also public to the world. These URLs are a true case of "security by obscurity", and by putting these URLs in standard public place all obscurity is removed. ☹️

**Comment by lucadgtrip** [comment]
The arguments in favour of Exporting iCal too are:
- uniformity: I know of no booking platform which only Imports, so the expected behaviour and capability is that both Import and Export are available. In other words not having Export will rise eyebrows, generate questions, support and so on.
- It encourages users to have Origin as the Main Calendar from which all the others are automatically synced. 
The main Calendar is the one you keep constantly an eye on, come back often to and improves the chances that Origin users will consider Origin as their main site.
Hosts will come back everyday and this is extremely precious.
That in turn will make it easier for them to update prices, listing details and so on. Having users constantly on the site opens a whole new set of possibilities.
You don’t want Hosts to come back just when a request comes in, find an unfamiliar UX, maybe retrieve the lost password and so on. I guess fidelization is what I’m referring to.
iCal Export capabilities is a pre-condition for fidelization, it does not guarantee it but it’s a first step.
It’s out of scope but there is a great opportunity in providing Hosts, especially the small ones ignored by paid Channel Managers, a basic and free channel management tool.
We may discuss this later on.
- It requires no gas, as other sites will read it.

The only arguments against exporting are:
- at least one more decision has to be taken: will Origin iCal export also bookings (blocked dates) imported from other iCals? Airbnb does not (or didn’t do until recently, I did’t check) and this generates confusion and overbookings. An example: you have two blocked dates in your Airbnb calendar:
10-12 August > booking generated in Airbnb.
12-14 August > booking generated in VRBO, and imported in Airbnb via Vrbo calendar.
the second one is NOT exportd by Airbnb.
So if you are using Airbnb as your main calendar and export it to, say, Booking.com, in Booking.com the dates 12-14 August are free, and an overbooking is generated.
- Responsibility: Hosts using Origin as their main calendar will blame Origin if something wrong happens (wrong reading, outage, the other site does not import ical well or on time etc..), but this is the same problem every platform has anyway.

>  If tons of people start using Origin dapps, it could look like a DDOS attack

if you let Dapps ping you have no control on how often they do it and it may be too often.
If you don’t let them chose, so Dapps may not be happy with the interval you chose and it centralizes the decision.
Anyway iCals are constantly pinged by all platforms, with many users such as Booking.com or Tripadvsor, Origin would be no different.
But Origin is different and the platforms may at a certain point boycott (openly or not) Origin iCal calls, so having this done by hundreds of Dapps may mitigate the problem as it would (maybe) be much more difficult for them to knows they are decentralised competitors.

> Josh's AirBnB iCal, because it contained guests' names and phone numbers

I see no reason to store nothing more than the dates. Maybe the Airbnb iCals could be sanitised and stripped of the extra data?

**Comment by wanderingstan** [comment]
@lucadgtrip writes:

> I see no reason to store nothing more than the dates. Maybe the Airbnb iCals could be sanitized and stripped of the extra data?

The trick is that there is no server to do the stripping and sanitizing. Of course we could do it on the "necessary evil" bridge server, but I loath to put still functionality into it.

**Comment by micahalcorn** [comment]
I am going to consider this done since we have a well-baked design document, are iterating on the API, and have real progress on the UI front. We should have some more specific DApp issues for the remaining work.
